### ISSUE TITLE
Image Upload and Storage

### ISSUE THREAD

**Issue by jacchau** [bloom-housing/bloom#129]
As we move from Listings being stored in a database and entered by users rather than stored in JSON and edited by engineers, we're going to need a more user-facing and scalable way to upload and manage image assets for the Listings.

Eventually, this will potentially expand to allow for other image assets, both multiple assets per listing (e.g. inside and outside views of a property) and then some jurisdiction-level assets once we're ready to have those be admin-user-editable (e.g. seal logos, banners), but in the short term we should be able to start with one image per listing.

In the further future, we might want to use a similar flow to upload and store applicant documentation like pay stubs or other proof of eligibility. This will add *significant* requirements around security, audibility, and compliance, since we'll be going from storing essentially public files (where the main risk is that someone could find an eventually-public image too early or before it's been properly approved) to storing very sensitive PII (where the downside of an information leak is a pretty large disaster). Ideally whatever we pick for a platform now will be able to support the more secure use-cases well, even if implementation is unlikely before Q42020.

Unless we uncover a specific reason it won't meet our needs, I think we're better off using a third-party service like Filestack, Cloudinary, Bynder, Uploadcare, but I think we're still open if there are compelling benefits to doing something ourselves, especially if there's no path to getting a reasonably secure solution.

**Comment by bk3c** [comment]
Also note that we want all of the version history and auditing contemplated in #286 available for image assets as well. This doesn't specifically need to be built in to whatever platform we choose -- we can version all of the metadata on our side and then just store each version of the asset separately, but we need to be able to see the changes over time.

**Comment by jaredcwhite** [comment]
I don't have experience with most of the options mentioned above. I have used Cloudinary on nearly every project I've touched over the last few years, so that would get a big upvote from me. It's very easy to generate size transformations on the fly (both in pixels and in kB), and by default it appends unique IDs so somebody could upload the same image filename over and over again and each one would be stored in a unique way (great for the versioned metadata use case). Their free plans are also very generous, so if each jurisdiction set up a separate Cloudinary account, it's probably free forever. :)
